core/src/main/scala/cats/syntax/either.scala

```diff
@@ -24,7 +24,7 @@ object EitherSyntax {
    * Uses the [[http://typelevel.org/cats/guidelines.html#partially-applied-type-params Partially Applied Type Params technique]] for ergonomics.
    */
   private[syntax] final class CatchOnlyPartiallyApplied[T](val dummy: Boolean = true ) extends AnyVal {
-    def apply[A](f: => A)(implicit CT: ClassTag[T], NT: NotNull[T]): Either[T, A] =
+    def apply[A](f: => A)(implicit CT: ClassTag[T]): Either[T, A] =
```

Comment:
While the `NT` parameter isn't referenced at runtime, it serves an important compile-time role. Consider that you wanted to do something like the following:

```scala
scala> Either.catchOnly[NumberFormatException]("foo".toInt)
res2: Either[NumberFormatException,Int] = Left(java.lang.NumberFormatException: For input string: "foo")
```

If you forget to add the type parameter before this PR you get a compile-time error:

```scala
scala> Either.catchOnly("foo".toInt)
<console>:18: error: ambiguous implicit values:
 both method If you are seeing this, you probably need to add an explicit type parameter somewhere, because Null is being inferred. in object NotNull of type => cats.NotNull[Null]
 and method catsAmbiguousNotNullNull2 in object NotNull of type => cats.NotNull[Null]
 match expected type cats.NotNull[Null]
       Either.catchOnly("foo".toInt)
```

However, with this PR you get a runtime error because `Null` is inferred as the type parameter:

```scala
scala> Either.catchOnly("foo".toInt)
java.lang.NumberFormatException: For input string: "foo"
  at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
  at java.lang.Integer.parseInt(Integer.java:580)
  at java.lang.Integer.parseInt(Integer.java:615)
  at scala.collection.immutable.StringLike.toInt(StringLike.scala:301)
  at scala.collection.immutable.StringLike.toInt$(StringLike.scala:301)
  at scala.collection.immutable.StringOps.toInt(StringOps.scala:29)
  at .$anonfun$res0$1(<console>:15)
  at scala.runtime.java8.JFunction0$mcI$sp.apply(JFunction0$mcI$sp.java:12)
  at cats.syntax.EitherSyntax$CatchOnlyPartiallyApplied$.apply$extension(either.scala:29)
  ... 39 elided
```


I think that the correct change to make here is to document why this `NotNull` parameter exists. The same goes for the `Validated` case.